lib/promql/expr.go

```diff
@@ -0,0 +1,423 @@
+package promql
+
+import (
+	"fmt"
+	"strconv"
+)
+
+// An Expr represents a parsed Extended PromQL expression
+type Expr interface {
+	// AppendString appends string representation of expr to dst.
+	AppendString(dst []byte) []byte
+}
+
+// A StringExpr represents a string
+type StringExpr struct {
+	S string
+}
+
+// AppendString appends string representation of expr to dst.
+func (se *StringExpr) AppendString(dst []byte) []byte {
+	return strconv.AppendQuote(dst, se.S)
+}
+
+// A StringTemplateExpr represents a string prior to applying a With clause
+type StringTemplateExpr struct {
```

Comment:
Instead of having a single node that can be in two states (in the past `StringExpr` had `tokens` and `S`, and depending on the whether the node had been WITH expanded one or the other was populated) I split into a separate node type, and applying the WITH clause changes this to a `StringExpr`